### PR TITLE
enable ASAN/UBSAN in pandas CI

### LINKED ISSUE #52990: ENH: Setup ASAN in CI
### Feature Type

- [ ] Adding new functionality to pandas

- [ ] Changing existing functionality in pandas

- [ ] Removing existing functionality in pandas

### Problem Description

We have no way of automatically detecting memory leaks in our library

### Feature Description

ASAN can be used to detect memory leaks with a relatively low cost. The following code can show how to build pandas with ASAN and find leaks. Note that `fast_unwind_on_malloc=0` severely slows down execution at the cost of a more informative traceback. This also reports leaks that likely cannot be controlled by pandas.

```sh
CC=clang CXX=clang++ CFLAGS="-fsanitize=address -fno-omit-frame-pointer -shared-libasan" LDSHARED="clang -shared" python setup.py build_ext --inplace -j8 --with-debugging-symbols
ASAN_OPTIONS="fast_unwind_on_malloc=0" LD_PRELOAD=$(clang -print-file-name=libclang_rt.asan-x86_64.so) python -m pytest pandas/tests/io/json/ 2> leaks.txt
```

This produces the following output, which contains items like:

```sh
...

Indirect leak of 24 byte(s) in 6 object(s) allocated from:
    #0 0x7fa3374d2c7e in __interceptor_malloc build-llvm/tools/clang/stage2-bins/runtimes/runtimes-bins/compiler-rt/lib/asan/asan_malloc_linux.cpp:69:3
    #1 0x7fa2c96590e4 in traced_malloc /home/willayd/clones/pandas/pandas/_libs/src/klib/khash_python.h:27:18
    #2 0x7fa2c96c793f in kh_resize_uint64 /home/willayd/clones/pandas/pandas/_libs/src/klib/khash.h:712:1
    #3 0x7fa2c96e80d3 in __pyx_pf_6pandas_5_libs_9hashtable_15UInt64HashTable___cinit__ /home/willayd/clones/pandas/pandas/_libs/hashtable.c:33343:3
    #4 0x7fa2c96e7861 in __pyx_pw_6pandas_5_libs_9hashtable_15UInt64HashTable_1__cinit__ /home/willayd/clones/pandas/pandas/_libs/hashtable.c:33275:13
    #5 0x7fa2c96c92f0 in __pyx_tp_new_6pandas_5_libs_9hashtable_UInt64HashTable /home/willayd/clones/pandas/pandas/_libs/hashtable.c:158300:7
    #6 0x55f9503681f5 in type_call /usr/local/src/conda/python-3.11.3/Objects/typeobject.c:1100:11
    #7 0x7fa2c5703998 in __Pyx_PyObject_Call /home/willayd/clones/pandas/pandas/_libs/index.c:82754:14
    #8 0x7fa2c5739d3b in __Pyx__PyObject_CallOneArg /home/willayd/clones/pandas/pandas/_libs/index.c:83070:14
    #9 0x7fa2c57370e9 in __Pyx_PyObject_CallOneArg /home/willayd/clones/pandas/pandas/_libs/index.c:83089:12
    #10 0x7fa2c571f6ab in __pyx_f_6pandas_5_libs_5index_12UInt64Engine__make_hash_table /home/willayd/clones/pandas/pandas/_libs/index.c:26319:89
    #11 0x7fa2c570fa8f in __pyx_f_6pandas_5_libs_5index_11IndexEngine__ensure_mapping_populated /home/willayd/clones/pandas/pandas/_libs/index.c:9532:17
    #12 0x7fa2c570cd7d in __pyx_f_6pandas_5_libs_5index_11IndexEngine__do_unique_check /home/willayd/clones/pandas/pandas/_libs/index.c:8653:15
    #13 0x7fa2c575e77a in __pyx_pf_6pandas_5_libs_5index_11IndexEngine_9is_unique___get__ /home/willayd/clones/pandas/pandas/_libs/index.c:8583:17
    #14 0x7fa2c575e67c in __pyx_pw_6pandas_5_libs_5index_11IndexEngine_9is_unique_1__get__ /home/willayd/clones/pandas/pandas/_libs/index.c:8549:13
    #15 0x7fa2c575e578 in __pyx_getprop_6pandas_5_libs_5index_11IndexEngine_is_unique /home/willayd/clones/pandas/pandas/_libs/index.c:75082:10
    #16 0x55f95038854d in _PyObject_GenericGetAttrWithDict /usr/local/src/conda/python-3.11.3/Objects/object.c:1278:19
    #17 0x55f95036adcf in PyObject_GenericGetAttr /usr/local/src/conda/python-3.11.3/Objects/object.c:1368:12
    #18 0x55f95036adcf in PyObject_GetAttr /usr/local/src/conda/python-3.11.3/Objects/object.c:916:19
    #19 0x55f950374b61 in _PyEval_EvalFrameDefault /usr/local/src/conda/python-3.11.3/Python/ceval.c:3465:29
    #20 0x55f95039acd2 in _PyEval_EvalFrame /usr/local/src/conda/python-3.11.3/Include/internal/pycore_ceval.h:73:16
    #21 0x55f95039acd2 in _PyEval_Vector /usr/local/src/conda/python-3.11.3/Python/ceval.c:6438:24
    #22 0x55f95039acd2 in _PyFunction_Vectorcall /usr/local/src/conda/python-3.11.3/Objects/call.c:393:16
    #23 0x7fa2c8db64d8 in __Pyx_PyObject_Call /home/willayd/clones/pandas/pandas/_libs/properties.c:5508:14
    #24 0x7fa2c8db684b in __Pyx__PyObject_CallOneArg /home/willayd/clones/pandas/pandas/_libs/properties.c:5576:14
    #25 0x7fa2c8db6309 in __Pyx_PyObject_CallOneArg /home/willayd/clones/pandas/pandas/_libs/properties.c:5595:12
    #26 0x7fa2c8db7f67 in __pyx_pf_6pandas_5_libs_10properties_14CachedProperty_2__get__ /home/willayd/clones/pandas/pandas/_libs/properties.c:1972:93
    #27 0x7fa2c8db6d2c in __pyx_pw_6pandas_5_libs_10properties_14CachedProperty_3__get__ /home/willayd/clones/pandas/pandas/_libs/properties.c:1723:13
    #28 0x7fa2c8db3b28 in __pyx_tp_descr_get_6pandas_5_libs_10properties_CachedProperty /home/willayd/clones/pandas/pandas/_libs/properties.c:4149:7
    #29 0x55f95038854d in _PyObject_GenericGetAttrWithDict /usr/local/src/conda/python-3.11.3/Objects/object.c:1278:19
    #30 0x55f95036adcf in PyObject_GenericGetAttr /usr/local/src/conda/python-3.11.3/Objects/object.c:1368:12
    #31 0x55f95036adcf in PyObject_GetAttr /usr/local/src/conda/python-3.11.3/Objects/object.c:916:19

SUMMARY: AddressSanitizer: 9878368 byte(s) leaked in 8924 allocation(s).
```

So from running the JSON tests ASAN thinks we leak 9878368 bytes with 24 being leaked by an interaction between properties.pyx -> index.pyx -> hashtable.pyx -> khash

Many of the arguments pieced together to make the above work are taken from:

google/sanitizers/issues/918
https://stackoverflow.com/questions/48833176/get-location-of-libasan-from-gcc-clang
https://clang.llvm.org/docs/AddressSanitizer.html

[leaks.txt](https://github.com/pandas-dev/pandas/files/11357716/leaks.txt)

### Alternative Solutions

Valgrind is another great tool for detecting memory leaks (amongst other things), but has larger overhead compared to the sanitizers, which *may* make it unsuitable for CI

https://github.com/google/sanitizers/wiki/AddressSanitizerComparisonOfMemoryTools

### Additional Context

cc @lithomas1

### PR DESCRIPTION
**Pull request by WillAyd**
@lithomas1 @rgommers 

closes https://github.com/pandas-dev/pandas/issues/52990

### CODE REVIEW

File: .github/actions/build_pandas/action.yml
Comment (WillAyd) [inline .github/actions/build_pandas/action.yml:28]: Probably don't want to hard code this - is there a way with GHA to only do this for certain action invocations @lithomas1 ?

File: .github/workflows/unit-tests.yml
Comment (WillAyd) [inline .github/workflows/unit-tests.yml:161]: Meson has the config option for `b_sanitize=address,undefined` but I don't think it set this. UBSAN has non-fatal errors, so without this things just get printed to stderr and pytest still continues.

Looks like NumPy does something similar with `halt_on_error=1`, but that didn't seem to stop pytest from continuing as I tried this locally

https://github.com/numpy/numpy/pull/24208/files

File: .github/actions/build_pandas/action.yml
Comment (mroeschke) [reply .github/actions/build_pandas/action.yml:28]: You can define an `input` above in the workflow and pass a variable from the job when you want to enable these flags

Comment (lithomas1) [reply .github/actions/build_pandas/action.yml:28]: I can push to your branch if you need any help with this, but it should be as Matt stated.

Comment (WillAyd) [reply .github/actions/build_pandas/action.yml:28]: Feel free to push

File: .github/workflows/unit-tests.yml
Comment (WillAyd) [inline .github/workflows/unit-tests.yml:159]: Having trouble setting this with GHA. There may also be a way to pass the flag directly through meson python? @lithomas1 any idea?

File: .github/workflows/unit-tests.yml
Comment (lithomas1) [inline .github/workflows/unit-tests.yml:167]: IIRC, this doesn't set CFLAGS inside of the action. It just makes ``env.CFLAGS`` available.

Why don't you try setting CFLAGs in action.yml if sanitize = true?

Comment (WillAyd) [reply .github/workflows/unit-tests.yml]: Ah great idea - much cleaner

**Comment by WillAyd** [comment]
Somewhat working now.  Guessing we need a way for pytest-xdist to fail and signal the test it failed on. Right now looks like a worker just crashed

**Comment by rgommers** [comment]
This looks pretty good to me. Let me Cc @ngoldbaum, who implemented the NumPy CI job and has more experience with these sanitizers than I have.

**Comment by ngoldbaum** [comment]
> Guessing we need a way for pytest-xdist to fail and signal the test it failed on.

Ah that's probably why the `halt_on_error=1` didn't work, I wasn't running with `pytest-xdist` locally when I was setting up the numpy config.

**Comment by mroeschke** [comment]
> Somewhat working now. Guessing we need a way for pytest-xdist to fail and signal the test it failed on. Right now looks like a worker just crashed

IMO I would add 1 new testing job in the matrix (e.g. that uses a 3.11 dependency file) that runs the tests with sanitize=True and python xdist with `-n 0`

**Comment by WillAyd** [comment]
In that case are you still planning to run against the entire test base or a subset of modules? I think removing multiple workers would slow down our CI a good deal? But maybe this gets to a state where it only runs when C/Cython files are touched?

**Comment by WillAyd** [comment]
Worth noting I tried `-n 0` locally a few times and it didn't make a difference. Not sure if the mere installation of pytest-xdist changes that. Needs further investigation

**Comment by lithomas1** [comment]
> In that case are you still planning to run against the entire test base or a subset of modules? I think removing multiple workers would slow down our CI a good deal? But maybe this gets to a state where it only runs when C/Cython files are touched?

We should do a minimal run, kind of like the npdev situation. So only us, numpy, and arrow installed.
Maybe that helps with the runtime?

**Comment by mroeschke** [comment]
Since the GHA Ubuntu runners only have 2 cores, I think running the entire test suite (even with all the dependencies) with `-n 0` will be _that_ significant, thought I don't know the impact that will have on the debugger.

FWIW that's how the Windows tests run currently and they take 15ish minutes longer than the non xdist runs

**Comment by WillAyd** [comment]
> thought I don't know the impact that will have on the debugger.

In theory the average runtime of ASAN would be 2x (see https://github.com/google/sanitizers/wiki/AddressSanitizer), though since we are not detecting leaks but also adding UBSAN I'm not sure how that all evens out

**Comment by WillAyd** [comment]
A lot of the datetime stuff in this PR is hacked together just to appear UBSAN, but there are definitely quite a few code paths where datetime conversions can lead to undefined behavior. 

The current ASAN failure looks like it comes from matplotlib, so @lithomas1 is probably right in that we need to pare this down to a smaller set of packages that we know can be clean

File: .github/actions/build_pandas/action.yml
Comment (WillAyd) [inline .github/actions/build_pandas/action.yml:35]: If we like this design we could also make `-e` a part of the meson_args input

File: .github/workflows/unit-tests.yml
Comment (WillAyd) [inline .github/workflows/unit-tests.yml:327]: Without this switch ASAN will in some cases recover and continue, which doesn't help CI to actually error out when issues occur (at least not with pytest-xdist)

**Comment by WillAyd** [comment]
Not entirely sure what the current CI issue is - @lithomas1 any chance you've seen this before?

ERROR: MISSING test execution (tx) nodes: please specify --tx

File: .github/actions/build_pandas/action.yml
Comment (rgommers) [reply .github/actions/build_pandas/action.yml:35]: That's mostly a `pip` argument rather than a meson/meson-python one. Probably better to keep `meson_args` for `config_settings` arguments only?

Comment (WillAyd) [reply .github/actions/build_pandas/action.yml:35]: Ah good point

**Comment by WillAyd** [comment]
Ok I think this is ready in its current form. Some things that can be improved in follow ups:

- more coverage. Runtime is good, but the full suite was timing out. A better balance can be found
- cleaning up some skips. Especially the address violations are culprits for segfaults
- Improving CI error messaging. Right now an error causes CI to fail, but the messaging back from a failed node through pytest-xdist needs improvement

**Comment by WillAyd** [comment]
Any feedback on this? I think would be nice to get in before the next release. Hope this eliminates the segfaults popping up during release

**Comment by lithomas1** [comment]
I'll look in a couple more days, if you still need my review.

Just finished finals, so gonna take a breather for a bit before doing pandas stuff.

File: .github/workflows/unit-tests.yml
Comment (mroeschke) [inline .github/workflows/unit-tests.yml:320]: Did you chose this file since you wanted a minimal environment? If so might as well create a new env file with a stable version of numpy

File: pyproject.toml
Comment (mroeschke) [inline pyproject.toml:527]: Could you rename this to something like `skip_asan: Tests known to fail ASAN check`?

File: .github/workflows/unit-tests.yml
Comment (mroeschke) [reply .github/workflows/unit-tests.yml:327]: If we get more predictable/consistent results without xdist, I think we should not run the tests in parallel

File: .github/workflows/unit-tests.yml
Comment (mroeschke) [inline .github/workflows/unit-tests.yml:308]: I think you might be able to fit this into the existing `jobs: ubuntu:` matrix

```
        include:
          - name: "ASAN_UBSAN"
            env_file: ci/deps/actions-311-numpydev.yaml
            pattern: "not slow and not network and not single_cpu and not skip_asan"
            asan_options: "detect_leaks=0"
            meson_args: --config-settings=setup-args="-Db_sanitize=address,undefined"
            ...
```

Then fields will be accessible with `matrix.field` and will be empty by default

File: pyproject.toml
Comment (WillAyd) [reply pyproject.toml]: Sounds good. One other thing I thought about though that is probably confusing for other developers - ASAN / UBSAN are different checks. So we could either have `skip_ubsan` and `skip_asan` as separate markers or maybe just a generic `skip_sanitizers` if we don't care that much about the distinction.

I'd have a slight preference to keep them separate - at least within our code base today ASAN violations cause segfaults, a lot of the UBSAN violations are tucked away in datetime extensions and _usually_ get covered up by other methods, so they are arguably less serious for pandas today. 

...but if others find that distinction confusing or unnecessary its not the end of the world to just have one marker

**Review by mroeschke — APPROVED** [review]
LGTM. So if there is a memory leak the new job is expected to fail?

**Comment by WillAyd** [comment]
Ah not yet unfortunately. The scope of this as is detects undefined behavior and address violations. ASAN can normally check memory leaks for you, but we had to add `detect_leaks=0` as an argument to the environment because the Python runtime seems to show all types of leaks

Whether those are legitimate or not requires more investigation. We could try to include a leak suppression file as described here https://github.com/pandas-dev/pandas/issues/54865#issuecomment-1839955886 but some of those suppressions also cast a very wide net to the point that it may not be all that useful

**Comment by mroeschke** [comment]
Gotcha. Or rather, if we introduce undefined behavior and address violations, this job will hopefully fail correct? Just want to ensure there's a definitive `job failure -> rectification -> job success` path for this job

**Comment by WillAyd** [comment]
Yes exactly - this will fail when either of those are detected

**Comment by WillAyd** [comment]
The error messaging you see in CI is something that could be improved. It just "fails" right now but that feedback gets lost along the way from the crashed process. I think that can be tackled in a follow up

**Comment by WillAyd** [comment]
This is what happens today if either of these pops up:

https://github.com/pandas-dev/pandas/actions/runs/7066657149/job/19241456914#step:8:61

**Comment by mroeschke** [comment]
> This is what happens today if either of these pops up:
> 
> https://github.com/pandas-dev/pandas/actions/runs/7066657149/job/19241456914#step:8:61

Ah OK. Could you at least include a `test_args: "-v"` in the job configuration in `unit-test.yml`? At least then the last test should be printed before the job fails so when it fails we don't have to rerun to figure out where this fails

**Comment by WillAyd** [comment]
OK sure. Here is what that looks like:

https://github.com/pandas-dev/pandas/actions/runs/7267349987/job/19800967532?pr=55102#step:8:18053

Ends up being too much for GHA to show but if you go to the raw logs you will see the error:

<details>

```
2023-12-19T21:11:17.6390164Z ../../pandas/_libs/src/vendored/ujson/python/objToJSON.c:2066:3: runtime error: signed integer overflow: 2147483647 + 1 cannot be represented in type 'int'
2023-12-19T21:12:44.8157270Z pandas/tests/io/test_common.py::TestCommonIOCapabilities::test_write_missing_parent_directory[to_json-os-OSError-json] 
2023-12-19T21:12:44.8158122Z [gw3] node down: Not properly terminated
2023-12-19T21:12:44.8158390Z 
2023-12-19T21:12:44.8158505Z replacing crashed worker gw3
2023-12-19T21:12:44.8158882Z INTERNALERROR> Traceback (most recent call last):
2023-12-19T21:12:44.8159868Z INTERNALERROR>   File "/home/runner/micromamba/envs/test/lib/python3.11/site-packages/_pytest/main.py", line 271, in wrap_session
2023-12-19T21:12:44.8160796Z INTERNALERROR>     session.exitstatus = doit(config, session) or 0
2023-12-19T21:12:44.8161413Z INTERNALERROR>                          ^^^^^^^^^^^^^^^^^^^^^
2023-12-19T21:12:44.8162455Z INTERNALERROR>   File "/home/runner/micromamba/envs/test/lib/python3.11/site-packages/_pytest/main.py", line 325, in _main
2023-12-19T21:12:44.8163512Z INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
2023-12-19T21:12:44.8164538Z INTERNALERROR>   File "/home/runner/micromamba/envs/test/lib/python3.11/site-packages/pluggy/_hooks.py", line 493, in __call__
2023-12-19T21:12:44.8165536Z INTERNALERROR>     return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
2023-12-19T21:12:44.8166476Z INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-12-19T21:12:44.8167481Z INTERNALERROR>   File "/home/runner/micromamba/envs/test/lib/python3.11/site-packages/pluggy/_manager.py", line 115, in _hookexec
2023-12-19T21:12:44.8168482Z INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
2023-12-19T21:12:44.8169138Z INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-12-19T21:12:44.8170131Z INTERNALERROR>   File "/home/runner/micromamba/envs/test/lib/python3.11/site-packages/pluggy/_callers.py", line 152, in _multicall
2023-12-19T21:12:44.8170977Z INTERNALERROR>     return outcome.get_result()
2023-12-19T21:12:44.8171378Z INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^
2023-12-19T21:12:44.8172287Z INTERNALERROR>   File "/home/runner/micromamba/envs/test/lib/python3.11/site-packages/pluggy/_result.py", line 114, in get_result
2023-12-19T21:12:44.8173179Z INTERNALERROR>     raise exc.with_traceback(exc.__traceback__)
2023-12-19T21:12:44.8174405Z INTERNALERROR>   File "/home/runner/micromamba/envs/test/lib/python3.11/site-packages/pluggy/_callers.py", line 77, in _multicall
2023-12-19T21:12:44.8175220Z INTERNALERROR>     res = hook_impl.function(*args)
2023-12-19T21:12:44.8175633Z INTERNALERROR>           ^^^^^^^^^^^^^^^^^^^^^^^^^
2023-12-19T21:12:44.8176555Z INTERNALERROR>   File "/home/runner/micromamba/envs/test/lib/python3.11/site-packages/xdist/dsession.py", line 123, in pytest_runtestloop
2023-12-19T21:12:44.8177342Z INTERNALERROR>     self.loop_once()
2023-12-19T21:12:44.8178296Z INTERNALERROR>   File "/home/runner/micromamba/envs/test/lib/python3.11/site-packages/xdist/dsession.py", line 148, in loop_once
2023-12-19T21:12:44.8179020Z INTERNALERROR>     call(**kwargs)
2023-12-19T21:12:44.8179889Z INTERNALERROR>   File "/home/runner/micromamba/envs/test/lib/python3.11/site-packages/xdist/dsession.py", line 273, in worker_collectionfinish
2023-12-19T21:12:44.8180703Z INTERNALERROR>     self.sched.schedule()
2023-12-19T21:12:44.8181588Z INTERNALERROR>   File "/home/runner/micromamba/envs/test/lib/python3.11/site-packages/xdist/scheduler/loadscope.py", line 339, in schedule
2023-12-19T21:12:44.8182381Z INTERNALERROR>     self._reschedule(node)
2023-12-19T21:12:44.8183278Z INTERNALERROR>   File "/home/runner/micromamba/envs/test/lib/python3.11/site-packages/xdist/scheduler/loadscope.py", line 321, in _reschedule
2023-12-19T21:12:44.8184099Z INTERNALERROR>     self._assign_work_unit(node)
2023-12-19T21:12:44.8185057Z INTERNALERROR>   File "/home/runner/micromamba/envs/test/lib/python3.11/site-packages/xdist/scheduler/loadscope.py", line 259, in _assign_work_unit
2023-12-19T21:12:44.8186005Z INTERNALERROR>     worker_collection = self.registered_collections[node]
2023-12-19T21:12:44.8186635Z INTERNALERROR>                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
2023-12-19T21:12:44.8187150Z INTERNALERROR> KeyError: <WorkerController gw7>
```

</details>

**Comment by mroeschke** [comment]
Great thanks!

**Comment by lithomas1** [comment]
> > This is what happens today if either of these pops up:
> > https://github.com/pandas-dev/pandas/actions/runs/7066657149/job/19241456914#step:8:61
> 
> Ah OK. Could you at least include a `test_args: "-v"` in the job configuration in `unit-test.yml`? At least then the last test should be printed before the job fails so when it fails we don't have to rerun to figure out where this fails

Hm, I wonder if there's a better way to do this.
(My main worry is that this make the logs very hard to scroll through, e.g. if I had to look through the logs because of a flaky non-ASAN related test)

Do you know if e.g. something like ``PYTEST_CURRENT_TEST`` might help?
https://docs.pytest.org/en/7.1.x/example/simple.html#pytest-current-test-env

**Comment by WillAyd** [comment]
Is there anything particular to CI that we know of that does not redirect stderr from pytest-xdist to the logs? If you run things locally you get the error and a stacktrace. While that isn't 1-to-1 to the test name it is pretty helpful to figure out what is going on so might be the best medium

**Comment by mroeschke** [comment]
> Is there anything particular to CI that we know of that does not redirect stderr from pytest-xdist to the logs? 

If I understand your question, this is a pytest xdist limitation: https://pytest-xdist.readthedocs.io/en/stable/known-limitations.html#output-stdout-and-stderr-from-workers

I know the `-v` isn't elegant, but I would prefer to have _a_ way to narrow down what is causing the failure before merging this in.

**Comment by lithomas1** [comment]
> I know the `-v` isn't elegant, but I would prefer to have _a_ way to narrow down what is causing the failure before merging this in.

I agree. I don't mean to block this PR, but maybe we should look into the reporting a little more?

At the very least, I think I could settle for a solution where we get pytest to print the filenames of the tests.
We could have this if we turned off pytest-xdist.
(there seems to be an issue with pytest-xdist where it swallows the filenames.
https://github.com/pytest-dev/pytest-xdist/issues/450)

Would this work?

**Comment by WillAyd** [comment]
OK here is what @lithomas1 suggestion looks like:

https://github.com/pandas-dev/pandas/actions/runs/7281808581/job/19843041899?pr=55102#step:8:590

Out of the two options so far I would prefer to go that route. It looks like turning off pytest-xdist for the ASAN build had little to no effect on the overall runtime

File: pyproject.toml
Comment (mroeschke) [inline pyproject.toml:526]: This one was removed recently

**Comment by lithomas1** [comment]
> OK here is what @lithomas1 suggestion looks like:
> 
> https://github.com/pandas-dev/pandas/actions/runs/7281808581/job/19843041899?pr=55102#step:8:590
> 
> Out of the two options so far I would prefer to go that route. It looks like turning off pytest-xdist for the ASAN build had little to no effect on the overall runtime

Ok, this looks correct to me, at a first glance. Just to double check, the failing test happens in ``test_common.py``, right?
(and not in ``pandas/tests/extension/json/test_json.py``)

**Comment by WillAyd** [comment]
Yea it does happen in test_common. I suppose the downside to this one is you don't get exactly the test that failed, but the first one I see failing locally is `pandas/tests/io/test_common.py::TestCommonIOCapabilities::test_write_missing_parent_directory[to_json-os-OSError-json]`

**Review by lithomas1 — APPROVED** [review]
LGTM (pending resolution of Matt's last comment)!

Excited to see this finally go in.

File: pyproject.toml
Comment (WillAyd) [reply pyproject.toml:526]: Ah sorry mistake on merge - nice catch

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Awesome! Thanks @WillAyd

**Pull request merged**
